### PR TITLE
Hcf backend

### PR DESCRIPTION
**Pull request by kalessin**

### CODE REVIEW

**Comment by sibiryakov** [comment]
Guys, sorry for stupid question. What would be the benefit for the open source community from HCF backend?

**Comment by kmike** [comment]
I agree with @sibiryakov. The problem with HCF backend in the same repo as crawl-frontier is that it is impossible to test for outside contributors. It means outside contributors can't refactor or change code easily if it touches HCF backend in any way. A separate package with HCF backend sounds good to me.

**Comment by kalessin** [comment]
I agree that the HCF backend cannot be part of an open source library. As for me, didn't know that this was the intention of this library. I will create a separate lib for this.

**Comment by kalessin** [comment]
There are a bunch of changes outside the hcf backend code itself, I will ask the people that did it, to split in a separate PR

File: tox.ini [hunk 1]
```diff
@@ -22,7 +22,7 @@ commands = flake8 setup.py crawlfrontier
 
 # Options for flake8
 [flake8]
-ignore = E265,E501,F401,W391,W292,E226
+ignore = E265,E501,F401,W391,W292,E226,E127
```
Comment (kalessin) [inline tox.ini]: @gatufo please make this change in a separate branch or in master

File: examples/scripts/10_custom_backends.py [hunk 2]
```diff
@@ -3,7 +3,8 @@
 """
 import random
 
-from crawlfrontier import FrontierManager, Settings, FrontierTester, graphs
+from crawlfrontier import FrontierManager, Settings, FrontierTester
+from crawlfrontier.utils import graphs
```
Comment (kalessin) [inline examples/scripts/10_custom_backends.py]: Separate unrelated changes

**Comment by kalessin** [comment]
@kmike @sibiryakov @gatufo I removed the hcf specific code from the branch, so here only left some changes needed in the core and memory backends needed for hcf, and some other changes that I cannot identify as required, but seems some code reorganization.

File: crawlfrontier/contrib/scrapy/schedulers/frontier.py [hunk 3]
```diff
@@ -78,7 +87,12 @@ def __init__(self, crawler):
         frontier_settings = crawler.settings.get('FRONTIER_SETTINGS', None)
         if not frontier_settings:
             log.msg('FRONTIER_SETTINGS not found! Using default frontier settings...', log.WARNING)
-        self.frontier = ScrapyFrontierManager(frontier_settings)
+
+        frontier_settings = Settings(frontier_settings or None)
+        frontier_settings.AUTO_START = False
```
Comment (sibiryakov) [inline crawlfrontier/contrib/scrapy/schedulers/frontier.py]: What's the motivation behind that?

File: crawlfrontier/core/manager.py [hunk 4]
```diff
@@ -324,13 +335,13 @@ def get_next_requests(self, max_next_requests=0):
                                             (max_next_requests, self.n_requests, self.max_requests or '-')))
 
         # get next requests
-        next_requests = self.backend.get_next_requests(max_next_requests)
+        next_requests = self.backend.get_next_requests(max_next_requests) or []
```
Comment (sibiryakov) [inline crawlfrontier/core/manager.py]: What's the reason for the type diversity here? May be it would be more readable if backend method get_next_requests will be allowed only to return list? Empty one, if there is no results.

File: crawlfrontier/contrib/scrapy/schedulers/frontier.py [hunk 5]
```diff
@@ -154,4 +170,24 @@ def _get_exception_code(self, exception):
...
 
     def _request_is_redirected(self, request):
-        return request.meta.get('redirect_times', 0) > 0
\ No newline at end of file
+        return request.meta.get('redirect_times', 0) > 0
+
+    def _add_middlewares(self, crawler):
+        """
+        Adds crawl-frontier scrapy scheduler downloader and spider middlewares.
+        Hack to avoid defining crawl-frontier scrapy middlewares in settings.
```
Comment (sibiryakov) [inline crawlfrontier/contrib/scrapy/schedulers/frontier.py]: IMO, there is no need to mention that it's a hack to avoid doing something in settings. I think the fact that you inject something automatically in Scrapy needs to be mentioned in documentation.

File: crawlfrontier/contrib/scrapy/schedulers/frontier.py [hunk 6]
```diff
@@ -120,11 +134,13 @@ def process_exception(self, request, exception, spider):
     def open(self, spider):
         log.msg('Starting frontier', log.INFO)
         if not self.frontier.manager.auto_start:
-            self.frontier.start()
+            scrapy_kwargs = {'spider': spider}
+            self.frontier.start(**scrapy_kwargs)
```
Comment (sibiryakov) [inline crawlfrontier/contrib/scrapy/schedulers/frontier.py]: Why not to write it that way: 
```python
self.frontier.start(spider=spider)
```
and in other places also?

**Comment by sibiryakov** [comment]
Guys, docs and tests needs to be updated also.

**Comment by kalessin** [comment]
rebased against development branch

**Comment by gatufo** [comment]
I'm closing this PR, to start a new clean one with hcf_backend changes rebased from master.
